exp/controllers/gcpmachinepool_controller.go

```diff
@@ -0,0 +1,330 @@
+		log.Error(err, "Error listing instances in instanceGroupManager")
+		return ctrl.Result{}, err
+	}
+
+	providerIDList := make([]string, len(igmInstances))
+
+	for i, instance := range igmInstances {
+		var providerID string
+
+		// Convert instance URL to providerID format
+		u := instance.Instance
+		u = strings.TrimPrefix(u, "https://www.googleapis.com/compute/v1/")
+		tokens := strings.Split(u, "/")
+		if len(tokens) == 6 && tokens[0] == "projects" && tokens[2] == "zones" && tokens[4] == "instances" {
+			providerID = fmt.Sprintf("gce://%s/%s/%s", tokens[1], tokens[3], tokens[5])
+		} else {
+			return ctrl.Result{}, fmt.Errorf("unexpected instance URL format: %s", instance.Instance)
+		}
+
+		providerIDList[i] = providerID
+	}
+
+	// FUTURE: do we need to verify that the instances are actually running?
+	machinePoolScope.GCPMachinePool.Spec.ProviderIDList = providerIDList
+	machinePoolScope.GCPMachinePool.Status.Replicas = int32(len(providerIDList))
+	machinePoolScope.GCPMachinePool.Status.Ready = true
+
+	// Requeue so that we can keep the spec.providerIDList and status in sync with the MIG.
+	// This is important for scaling up and down, as the CAPI MachinePool controller relies on
+	// the providerIDList to determine which machines belong to the MachinePool.
+	return ctrl.Result{RequeueAfter: 1 * time.Minute}, nil
+}
+
+func (r *GCPMachinePoolReconciler) reconcileDelete(ctx context.Context, machinePoolScope *scope.MachinePoolScope) error {
+	log := log.FromContext(ctx)
+
+	log.Info("Handling deleted GCPMachinePool")
+
+	if err := instancegroupmanagers.New(machinePoolScope).Delete(ctx); err != nil {
+		log.Error(err, "Error deleting instanceGroupManager")
+		r.Recorder.Eventf(machinePoolScope.GCPMachinePool, corev1.EventTypeWarning, "FailedDelete", "Failed to delete instancegroupmanager: %v", err)
+
+		// record.Warnf(machineScope.GCPMachine, "GCPMachineReconcile", "Reconcile error - %v", err)
+		conditions.MarkUnknown(machinePoolScope.GCPMachinePool, expinfrav1.MIGReadyCondition, expinfrav1.MIGNotFoundReason, "%s", err.Error())
+		return err
+	}
+
+	if err := instancetemplates.New(machinePoolScope).Delete(ctx); err != nil {
+		log.Error(err, "Error deleting instanceTemplates")
+		r.Recorder.Eventf(machinePoolScope.GCPMachinePool, corev1.EventTypeWarning, "FailedDelete", "Failed to delete instance template: %v", err)
+
+		// record.Warnf(machineScope.GCPMachine, "GCPMachineReconcile", "Reconcile error - %v", err)
+		conditions.MarkUnknown(machinePoolScope.GCPMachinePool, expinfrav1.InstanceTemplateReadyCondition, expinfrav1.InstanceTemplateReconcileFailedReason, "%s", err.Error())
+		return err
+	}
+
+	// remove finalizer
+	controllerutil.RemoveFinalizer(machinePoolScope.GCPMachinePool, expinfrav1.MachinePoolFinalizer)
+
+	return nil
+}
+
+func (r *GCPMachinePoolReconciler) getInfraCluster(ctx context.Context, cluster *clusterv1.Cluster, gcpMachinePool *expinfrav1.GCPMachinePool) (*infrav1.GCPCluster, *scope.ClusterScope, error) {
```

Comment:
```suggestion
func (r *GCPMachinePoolReconciler) createScope(ctx context.Context, cluster *clusterv1.Cluster, gcpMachinePool *expinfrav1.GCPMachinePool) (*scope.ClusterScope, error) {
```

Should this be createScope or so? We never use the InfraCluster returned from this func.